### ISSUE TITLE
Background map is incredibly old

### ISSUE THREAD

**Issue by mfbehrens99** [westnordost/StreetComplete#1700]
I am using Street Complete in an area where I added all the ways and roads some 5 Months ago however they still don't show up an the map in the background. The quests the app is showing me are fine. How often does the map get updated? Or is there maybe a bug. If not I would suggest to update the map data more frequent.

**Comment by Akasch** [comment]
At the moment I don't have the server capacities to host the whole World. Therefore some parts are served by https://www.nextzen.org/, but as far as I know the maps from them are older. 

If you tell me which region you are from, I can see if it is possible to include it into my server (or you can sponsor a server or host the region yourselve)

**Comment by westnordost** [comment]
That most of the world ([outside of the area highlighted in green here](https://map-data.de/#map=2.75/14.086925694319826/17.64761611554013&style=cinnabar&show_server)) is very old is a known problem, on the upside, nextzen is provided free.

Though since the up-to-dateness is a problem for many people, I could start thinking again about using a commercial vector tile hosting service and/or ask them if they'd like to sponsor this app by granting their service for free. Though right now I am busy with other urgent things regarding this app.

There are a number of other vector tile hosters around for who it must be ascertained first if their tiles are (a) regularly updated worldwide, (b) no/very few downtimes during tile re-generation and (c) are compatible with tangram-es. [switch2osm](https://switch2osm.org/providers/) lists a few, of which it looks like these offer vector tile hosting:
- https://www.thunderforest.com/
- https://stadiamaps.com/
- https://www.maptiler.com
- https://www.mapbox.com/
- https://www.jawg.io

At least the first two are from people which are active in the OSM community which is a plus. So anyone wanting to do some nonbinding enquiries about the above mentioned requirements, please feel invited to do so! For Point (c), it must be MVT format of course, but the details are more difficult to acquire. For starters, it would help to ask if they serve the identical data as mapbox (or another big player) or something custom with hopefully a documentation.

Another angle would be to find someone, like FOSSGIS, who would sponsor the costs for a dedicated vector tile hosting server which @Akasch could maintain, if he is still up to it.

**Comment by ftsell** [comment]
Hi,
I'm a friend of @Akasch and currently working myself into what he has built as server infrastructure. I'm planning to do some major code cleanup and general hosting improvements in the course of this year. This will hopefully make everything more scalable, maintainable and up-to-date. It would also make it easier for others to chime in and help with hosting resources.

Of course we wouldn't say no to additional hosting resources and gladly administer them ;)

**Comment by westnordost** [comment]
That's nice to hear!
What would make it easier for others to help with hosting resources would be a dead-simple one-click setup. So, probably a ready-made docker container these days. Though if it is just python scripts(?), shoving a package of preconfigured scripts into a certain directory of a good ol' webspace might do the trick too.
Finally, how to host resources could be included in the CONTRIBUTING.md

I don't expect that much help in hosting though. Not many users who read this tracker will have rented some (v)server with free capacities and commitment to maintain it. Right now, Oceania is hosted by someone else than @Akasch right? Who was it again? This information could perhaps be added to [the map showing the servers](https://map-data.de/#map=2.75/14.086925694319826/17.64761611554013&style=cinnabar&show_server)?

**Comment by goldfndr** [comment]
Esri's OpenStreetMap vector basemap [is now in general release](https://www.esri.com/arcgis-blog/products/arcgis-online/mapping/whats-new-with-openstreetmap-basemap-march-2019/) and [there is a variety of basemaps](https://www.arcgis.com/home/group.html?id=5e0d56e27d7e4a68955d399ca96c41cb#overview).

There's a sample web deployment [via Tangram](https://www.jonahadkins.com/esri/cartography/openstreetmap/2019/03/28/esri-tangrams.html). Android deployment *might* require [the ArcGIS Runtime SDK](https://developers.arcgis.com/android/10-2/guide/use-arcgis-basemaps.htm) which presumably involves [specific licensing](https://developers.arcgis.com/android/10-2/guide/license-your-app.htm).

(Apologies if I've mentioned it previously; I haven't been keeping up these days.)

**Comment by lseelenbinder** [comment]
Hi @westnordost, I'm co-founder at Stadia Maps! I'm happy to answer any direct questions you have.

We use OpenMapTiles for our schema, with a few (currently only additive) adjustments. Our data update cadence may be problematic for your needs, but we can discuss that.

We'd be willing to find some middle-ground on costs since StreetComplete is so important to the ecosystem. Happy to talk further on here or via email: luke.seelenbinder@stadiamaps.com.

**Comment by westnordost** [comment]
Hi @lseelenbinder, thank you very much for the offer! I also met Ian on the SotM but I fear he might be a bit disappointed at me for not following up on the default speed limits research we've been doing a together a bit. As said I am currently very busy with other things so I won't be able follow up on this for some time. The information about OpenMapTiles is valuable, I assume this format is used by several vector tile hosting services. It is left to find out if what of https://github.com/ENT8R/streetcomplete-mapstyle must be changed for the style to work with that format.

**Comment by lseelenbinder** [comment]
I think @ianthetechie understands—we're all insanely busy. 😄 

He also is a bit better at styles than I am, so I'll let him comment on how much work would be necessary to adjust the theme to our schema.

**Comment by ianthetechie** [comment]
No worries @westnordost haha. I've been equally busy.

Regarding the style, tangram and Mapbox GL JSON styles are a bit different, so you would have to do some work to restyle this, but since your style is pretty minimalist, I don't think this would be too much work.

**Comment by RubenKelevra** [comment]
Hey guys!

What about using a Collaborative Clusters from IPFS to help the guy(s) which offers the map?

This way everyone can just set up a IPFS client and add a follower to the cluster and provide storage as well as bandwidth.

The Gateway from IPFS to HTTPS can be provided from [Cloudflare](https://cloudflare-ipfs.com).

This way the map is always fast and tiles are redundantly stored in many different places and there's much more bandwidth available.

Info about Collaborative Clusters:

https://cluster.ipfs.io/documentation/collaborative/


I'm running a Collaborative Cluster, feel free to ask me for details how to set it up and feed it with the data. :)

**Comment by ftsell** [comment]
@RubenKelevra thanks for the idea, it certainly sounds like a viable alternative and I quite like the IPFS project too :)

It might just not be the solution to our actual problem but I see that this is not really clear from outside. 
Our problem is not really storage or bandwidth per se but rather the fact that we do processing on the whole OSM world dataset before committing them to storage. In addition to that, the storage is not just a dumb blob storage but a postgresql database with a lot of indices (for zoom levels and such) which makes querying the data a bit complex too.

Now we might be able to work something out by using IPFS but I don't yet see how

**Comment by RubenKelevra** [comment]
When I understand your pipeline correctly, you're downloading the OSM-XML file, process it to a PostgreSQL.

If there's a request you're the querying the PostgreSQL for the data which is inside the boundaries of the tile, like roads, houses etc. to generate the SVG, right?

So your sever is basically on it's limit if someone is querying a lot of tiles on an area you haven't processed before, because those SVGs cannot be delivered by your nginx.

Do you know the option `proxy_cache_use_stale updating;` combined with `proxy_cache_background_update on;` from Nginx?

This basically tells nginx to push old files to the clients where the time of life has already expired, while nginx asks the backend now for a new version of the file in the background, to update it's cache.

So you can basically delay the update process, while the clients gets outdated tiles but these are served very fast.

When the client asks the next time if the tile is up to date (like on the next day) nginx will deliver the new file to the client, since the backend has refreshed the tile.

**Comment by RubenKelevra** [comment]
@ftsell 

Another option would be to switch to a different database. RethinkDB for example can handle geo data very well and can natively store points, lines and polygons.

The clustering of more than one server and sharding/replication of data can be done on the fly and very easy on a web interface.

Queries on a cluster is calculated in parallel by all shard members and send back to the client when completed in a transparent way, so the client doesn't have to care if he speaks to one server or a cluster. You can also use this to run stuff on multiple servers, since the dataset is writeable on all cluster members. :)

It supports native Python, Ruby, Java and JavaScript functions, so you don't have to create SQL queries, but write your queries in native functions which I found always pretty intuitive.

You can also import and export GeoJSON directly from and to the database.

A nice feature of RethinkDB is that you can stream updates to clients (called [changefeeds](https://rethinkdb.com/docs/changefeeds/python/)), so you can query the database and if there are updates which match the query, they are send to the client in real time.

RethinkDB also supports map & reduce which runs inside the database cluster.

Not sure if this would fit your needs, just a heads up that there are more options than just PostgreSQL :)

https://rethinkdb.com/docs/geo-support/python/

**Comment by rugk** [comment]
Generally said, would not it also be a first step to "open-source" the server config/"pipeline process". So people can see and directly create issues/comment on the setup/files like the nginx config etc.? (just as done above, but with guessing)

**Comment by Akasch** [comment]
@rugk most of the used code is in the repositories at https://github.com/Map-Data only the nginx configs for the distribution and caching are missing. But most of the code is not really documented. I will see if I can clean up the nginx configs and publish them to.

@RubenKelevra while I think IPFS and RethinkDB are both interresting projects I don't think we can use them to make the tileserver faster/better. One of the mayor points is that we would have to rewrite the used libraries, and even if we would update the Database based on the minutely osm changeset feeds, the problem is more with the preprocessing of the data, as to apply an update the data has to be preprocessed and correlated with existing data.

The problem at the moment is mostly that we do not have enough fast storage to make it possible to host nearly the full world again. Maybe some more CPU power is required, but I think with more SSD space most of the problems would mostly solvable.

I will try to write a document about the general structure and how the different parts are interacting with each other in the next days. @ftsell and I are also planning to make it possible to host a region via a simple Docker command, provided the server has ~ 50 GB free space (SSD is way faster in the initial setup) and ~4 CPU cores (some cores can be shared if more than one region is hosted on the same server)

**Comment by RubenKelevra** [comment]
> The problem at the moment is mostly that we do not have enough fast storage to make it possible to host nearly the full world again. 

@Akasch host what? Tiles or the database? 🤔

**Comment by RubenKelevra** [comment]
> @rugk most of the used code is in the repositories at https://github.com/Map-Data only the nginx configs for the distribution and caching are missing. But most of the code is not really documented. I will see if I can clean up the nginx configs and publish them to.
> 
> @RubenKelevra while I think IPFS and RethinkDB are both interresting projects I don't think we can use them to make the tileserver faster/better. One of the mayor points is that we would have to rewrite the used libraries, and even if we would update the Database based on the minutely osm changeset feeds, the problem is more with the preprocessing of the data, as to apply an update the data has to be preprocessed and correlated with existing data.

Yeah maybe you're right. It's interesting to get a basic overview to see how stuff can be improved. :)

**Comment by Akasch** [comment]
> @Akasch host what? Tiles or the database? thinking
Host the Databases to generate the tiles on request (or if one want pre generate tiles). 

To add to the requirements: while importing it is a good idea to have more than 5 GB of RAM, while running as regional tileserver, I think 2 GB should ok, but I havn't tested the RAM requirements.

**Comment by smichel17** [comment]
I have a pretty beefy coding/gaming desktop that's currently sitting around unused for most of the day. I don't have a good way to host from it, but I'm happy to donate compute power if possible.

**Comment by davidak** [comment]
@smichel17 this is unrelated, but have you considered to donate compute power to science using [BOINC](https://boinc.berkeley.edu/), for example in the [World Community Grid](https://www.worldcommunitygrid.org/)? It only runs when the CPU/GPU is idling.

**Comment by Suburbanno** [comment]
![Screenshot_20200329-090832](https://user-images.githubusercontent.com/13632826/77848743-7120ca80-719d-11ea-9d80-212fe309b775.png)
![Screenshot_20200329-090815](https://user-images.githubusercontent.com/13632826/77848748-754ce800-719d-11ea-8e46-6198fd1a8e6b.png)

**Comment by RubenKelevra** [comment]
@Suburbanno outch.

**Comment by michaelblyons** [comment]
Yeah, I've been lurking this issue since January. I have a similar problem wherein I can't (for instance) do any building-related quests in my local area because the building traces in OSM are newer than the data Nextzen has.

**Comment by RubenKelevra** [comment]
@westnordost how about getting in touch with Andy Allan from Thunderforest?

Using their Tile-API for an OpenStreetMap-Editor may be free of charge - JOSM at least uses their service for Tiles for years.

Their tile-services at least are quite nice to look at and really fast. :)

**Comment by lseelenbinder** [comment]
The offer from @stadiamaps is still on the table, as well. We're happy to sponsor vector tiles (MVT) for this.

I think the primary issue with switching is that it requires reworking the themes, etc. It's not as simple as a drop-in replacement.

**Comment by Akasch** [comment]
Currently I have the data preprocessing running, and will reimport the region of Europe in the next 1-2 days. After that I will look how many additional regions I can host, while doing it I try to make the process better so it can run automated and also is easier to run the parts of the process.

Switching to an other existing provider would, of course make it easier for me, but I think it would require a requite of the styles, I don't know how compatible they are.

**Comment by RubenKelevra** [comment]
@lseelenbinder thanks for the offer, it's greatly appreciated to get support :)

Sure, it needs work, but the current situation is pretty bad tbh. No pressure @Akasch, but we need something reliable which doesn't stand on shoulders which are already too busy, you know? :)

The best option would be to have two services to choose from, so if one has an outage the user can just switch from one map to a different one.

Doesn't have to be equal in style, maybe one more car-centric and one bike/foot-centric? :)

**Comment by westnordost** [comment]
It's on my list :-)

But well, anybody wanting to help making a possible conversion to a commercial map tiles provider easier, can do so by either:

1. by researching what schema the providers here use https://github.com/westnordost/StreetComplete/issues/1700#issuecomment-574361700 - and if it is compatible to i.e. OpenMapTiles (seems to be semi-standard?)

2. by creating a fork/branch of https://github.com/ENT8R/streetcomplete-mapstyle that is compatible with an OpenMapTiles-compatible schema, because as we know, at least stadiamaps and OpenMapTiles would work with that. Once this exists, a switch could be done in minutes.
Playing around with the map style sheet also does not require coding knowledge because the style sheet is an own language anyway and well documented. Also, the vector tile format is actually pretty standardized, so there will probably not be big differences. The differences that might occur usually are that one vector map tile provider might for example name all minor roads "minor_road" and the other uses OSM inspired names for the features (living_street, residential,...), or that one provider names the layers differently than the other.

**Comment by RubenKelevra** [comment]
@westnordost just want clearify, you want to keep the change as minimal as possible, visually speaking - or is there room for changes in the process?

I know there have been put large efforts in this style, but I always thought it looks a bit too dull.

I would love to see a bit more distinctions in path/road details, to maybe catch some mapping errors on the way (like a path is a track and vise versa), while not overloading the users with details (like the basemap does sometimes).

I always liked the color scheme of opentopomap.org because of the high contrast, which helps to identify map details in the bright sunlight on a smartphone.

And the mapstyle is CC-BY-SA 3.0, so we're not 'copying' someone's commercial work, when we're taking cues from this repo:

https://github.com/der-stefan/OpenTopoMap

**Comment by westnordost** [comment]
Yes, there should be no visible change when switching between vector tile servers. Offering different themes for users to select is another feature (request) that does not belong in this thread.

**Comment by RubenKelevra** [comment]
Alright :)

**Comment by davidbaumann** [comment]
Interesting program: https://wiki.openstreetmap.org/wiki/Tiles@home

**Comment by HolgerJeromin** [comment]
@davidbaumann 
Have you actually read the page? :-)
This is prehistoric software and discontinued since 8 years.

**Comment by davidbaumann** [comment]
I've seen that, but I really like the approach.

**Comment by tuxayo** [comment]
@westnordost 
> Yes, there should be no visible change when switching between vector tile servers.

How will cache invalidation be managed? For non-tech-savvy users.

**Comment by westnordost** [comment]
The switch is not something the user will do. The switch is the "migrating" to another tileserver

**Comment by tuxayo** [comment]
I was reffering to updating the outdated tiles already cached in the devices running StreetComplete as of now.

**Comment by westnordost** [comment]
Well they would be dropped

**Comment by smichel17** [comment]
I've been thinking more seriously about hosting from home. Two questions:

1. (@ anyone with familiarity) What are the bandwidth requirements for a tile server? I'm likely to be bottlenecked by 10 Mbps upload.
2. @westnordost Would it still be useful even if I cannot guarantee high availability? Eg, use my server for North America if it's online, but fall back to a different service if not. Or would that not be worth the hassle? (Say, ~90% uptime instead of >99%)

**Comment by tuxayo** [comment]
> Well they would be dropped

Great, I didn't know there was already an invalidation mechanism :D

**Comment by westnordost** [comment]
It's a normal HTTP cache, so if the tiles come from a different URL, the cache is different.

**Comment by wibrt** [comment]
Using another storage engine is unlikely to solve the issue of the maintainers, as the bytes need to be stored. It seems to be a capacity problem

Another way of looking at this, possibly wrong, i'm not familiar with the source code:
* get the data directly from osm?
* get the data from osmand (if on the device) ?
** friendly on the download and storage usage the device owner
** i think most people that have this app will also have osmand
** thus not storing the same information twice
* get the data from another up to date osm open source tiling service?

**Comment by westnordost** [comment]
First two options won't work. Third option is what this thread was about the last months. If you or anyone want to help, see https://github.com/westnordost/StreetComplete/issues/1700#issuecomment-605702153

**Comment by davidbaumann** [comment]
+1 for a docker container.
I would even run that on my workstation over night, for example.
It would be important to have small parts so I can also just run it for 2-3 hours or maybe 12 over night.

**Comment by HolgerJeromin** [comment]
ref #122 for offline map discussions including osmand.

**Comment by Helium314** [comment]
Like @smichel17 I'm thinking about hosting from home
I have several questions about this (@Akasch?):
1. will this actually provide recent map data, or just another copy of the outdated tiles?
2. what do I need to do? Setup from https://github.com/Map-Data/regiontileserver and get some nginx configuration?
3. is my computer actually good enough? 4 GB memory, i5 4300U CPU, ~70 GB to spare on SSD
4. is 50 Mbit/s upload enough?

**Comment by smichel17** [comment]
Hm. Maybe instead of DDNS I could set up a load balancer of some sort, that would forward people to my or @Helium314's home PCs if they're online and nextzen otherwise. That way we wouldn't exceed capacity and it wouldn't be as big a deal if I had less uptime. However, this has some complications -- I would want to send redirects or something rather than proxying all traffic through my hosting provider; maybe at that point we should just use bittorrent :D -- and is a bigger project that I'm not sure I have time for right now.

**Comment by Atrate** [comment]
You mention Bittorrent jokingly, but a distrivuted P2P network updated from a could of central servers could work of there was a protocol that would support quick peer discovery and reliable incremental updates

**Comment by rugk** [comment]
Don't claiming or evaluating (whether) this is the solution, but let's throw https://ipfs.io/ into the room… it's made for p2p and a lot of data.

**Comment by ftsell** [comment]
@Helium314 

> Like @smichel17 I'm thinking about hosting from home
> I have several questions about this (@Akasch?):

Great :+1: we're planning to make it possible for self-hosters like you to contribute their hosts to our server-mesh (if you wanna call it that) but we're not quite there yet.

> 1. will this actually provide recent map data, or just another copy of the outdated tiles?

Yes it will provide recent map data.

> 2. what do I need to do? Setup from https://github.com/Map-Data/regiontileserver and get some nginx configuration?

Essentially yes since [regiontileserver](https://github.com/Map-Data/regiontileserver) is the server which actually hosts the data. It does so from a postgresql database so you'll need one as well.
Currently the script for importing into postgres is in this repo too but we might change that.

Additionally [osm-tile-data-extract](https://github.com/Map-Data/osm-tile-data-extract) is used to download upstream data from OSM and split it up into smaller chunks so that one tileserver can serve one such smaller chunk later.

Now because we don't use on huge machine for serving the whole world, we also have [tileserver-mapping](https://github.com/Map-Data/tileserver-mapping). This server knows which tileserver handles what chunk and tells nginx via [X-Accell Header](https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/) the correct tileserver.

> 3. is my computer actually good enough? 4 GB memory, i5 4300U CPU, ~70 GB to spare on SSD

To host one or two tileservers for small chunks of the map, probably but for the whole world definitely not. The whole OSM planet dump is currently around 50GB and you will need some working disk space to do the splitting into smaller chunks. 

> 4. is 50 Mbit/s upload enough?

I don't really know but if it's not used too much probably.

---

I'll write this down properly in some documentation later this weekend ;)

**Comment by smichel17** [comment]
Thanks @ftsell, that's a great overview. One piece I'm missing: how does the data get transformed from raw OSM to vector tile format? I was looking at openmaptiles.org the other day and it seems like this process is fairly compute-intensive. Is that part of the script for importing to postgres?

**Comment by ftsell** [comment]
It's currently in [regiontileserver](https://github.com/Map-Data/regiontileserver) in some bash script but it would better fit into [osm-tile-data-extract](https://github.com/Map-Data/osm-tile-data-extract).

Yes it is fairly intensive because a lot of indexes have to be generated. It would be nice to do the transformation directly after we split the chunks and then simply provide postgresql dumps for others to download.

**Comment by Helium314** [comment]
So we need to
* split up the planet into regions
* import region(s) to postgresql
* host a tileserver for one or more regions

As far as I understand it shouldn't be hard to distribute this work, e.g.
* someone splits up the planet and hosts regions (probably good bandwidth is useful)
* import is done on several computers (because it is slow), and postgresql dumps are uploaded to the data server above. The computers doing the transformation do not need to be used as tile servers. This would allow to distribute the work better, e.g. I could use my main computer.
* tile servers download/import postgresql dumps for some regions and serve tiles

Now I don't know much about how all this really works, especially the database stuff.
So is this actually possible?

**Comment by ftsell** [comment]
Yes that sums it up pretty well. It _just_ needs to be implemented where most work needs to be done too coordinate / trigger these schedules.

**Comment by goldfndr** [comment]
> The whole OSM planet dump is currently around 50GB

If this is only for the background map then presumably most of that data can be filtered out via [osmfilter](https://wiki.openstreetmap.org/wiki/Osmfilter) or something similar. Doing a cursory look at the [map styles](https://github.com/ENT8R/streetcomplete-mapstyle), it looks like the tags used are limited to:

 ```
name
addr:housenumber
oneway
natural
boundary
waterway
building
landuse
amenity
leisure
barrier
location
highway
railway
aerialway
route=ferry
route=piste
aeroway
```

(and probably further limited by more specific values, e.g. specific barriers)

But that seems incomplete, as somehow relative building heights are extruded.

**Comment by westnordost** [comment]
In fact, I evaluated a few commercial vector tile providers in the meantime, see the different branches on https://github.com/ENT8R/streetcomplete-mapstyle

### openmaptiles (maptiler)
- barriers such as walls, fences, hedges, city walls etc. missing are missing. [It sounds like they do not plan to add it into the main source](https://support.maptiler.com/i745-vector-tile-schema-missing-barriers).
- a property to recognize private roads is missing. But they find it relevant and [I added a feature request](https://github.com/openmaptiles/openmaptiles/issues/925).
- mini roundabouts, turning loops and turning circles are missing. But they find it relevant and [I added a feature request](https://github.com/openmaptiles/openmaptiles/issues/926).
- 3D buildings don't render as good as for nextzen tiles, there are some bugs
- mediocre appearance on medium zoom level (z8): small landcovers are not merged together but just left out
- maptiler tiles are updated weekly

### openmaptiles (stadiamaps)
same as for maptiler, only:
- additionally, housenumbers are missing. They do not plan to add it to the main source but maybe provide it as a second source in the future.
- stadiamaps tiles are updated monthly. There are plans to get down to bi-weekly or even weekly updates.

### thunderforest
- very good appearance on medium zoom level (z8,...): small landcovers are merged together and geometry is simplified very carefully (or not at all?)
- the tiles are very huge, 2-4 times the size of openmaptiles on high zooms, several MB on medium zooms (z8,...). They are working on further reducing the size.
- there is no height information for buildings and thus no 3D buildings at all. They are aiming at adding at least basic support for 3D buildings (so, no building parts).
- mini roundabouts, turning loops and turning circles are missing. It doesn't have a high priority for them to add it.
- the tiles are updated every few days

### jawg.io
- good appearance on medium zooms (z8)
- the tiles are very lightweight, having roughly half the size of openmaptiles tiles. On the other hand, jawg.io is the only service of those tested that serves tiles until z16 (others till z14), meaning that there are 20x as many tiles served
- many types of landuses are missing. They are looking into adding some of them.
- they have only basic 3D buildings support (no building parts). It doesn't have a high priority for them to add it.
- mini roundabouts, turning loops and turning circles are missing. It doesn't have a high priority for them to add it.
- node barriers such as gate, bollard, lift_gate, cycle_barrier, stile, block, toll_booth, kissing_gate, swing_gate, chain, cattle_grid etc. are missing. They'll think about it whether to include it or not.
- a property to recognize private roads is missing
- no filtering between service rails and other possible. They do not plan to add this info.
- city walls are mapped to fences. They are thining about whether to keep it as an own category.
- the tiles are updated daily

I've looked at the tile sizes, but not at what actual traffic is generated. Next step would be to look at that. Some properties currently in the stylesheet are not that important, such as mini roundabouts etc.
The things I deem important are:
- size
- up-to-dateness
- 3D buildings
- housenumbers

I haven't looked closer at mapbox tiles yet as I didn't find the TileJSON url. What I know is that they also update daily and that they serve tiles till z16.

**Comment by ZLima12** [comment]
I could possibly host a server for the New York area.

**Comment by ENT8R** [comment]
> ### Mapbox

The documentation of the Mapbox vector tiles seems to be at https://docs.mapbox.com/vector-tiles/reference/mapbox-streets-v8/ but as far as I understand, it is not allowed to use them in your app without using their own Mapbox Mobile SDK.

In their [service terms](https://www.mapbox.com/legal/service-terms) they state the following:
> 7. **Mobile SDKs**. 
> 7.1. **General Requirements**. 
You must use the Mobile SDKs as your exclusive means of accessing the Services in mobile applications. The Mobile SDKs will periodically send location and usage data to us, which we may use for the purpose of fixing bugs and errors, accounting and generating aggregated anonymized statistics. [...]

**Comment by westnordost** [comment]
> The documentation of the Mapbox vector tiles seems to be at https://docs.mapbox.com/vector-tiles/reference/mapbox-streets-v8/ but as far as I understand, it is not allowed to use them in your app without using their own Mapbox Mobile SDK.

I think this is a pricing policy. MapBox has special pricing for usage of their tiles for when using their mobile SDK: You pay by number of users, not by number of requests, as far as I know. There is also pricing by number of requests, so maybe that wouldn't be a problem.

**Comment by westnordost** [comment]
Okay, I looked at the total download size by example of zooming in from z0 to z16 to the city center of Hamburg. 

Relative to openmaptiles (stadiamaps),
- thunderforest is about **8 times** the size for zoom levels 1-12 and about **3 times** the size for zoom levels 13-16
- jawg.io is about **half** the size for zoom levels 1-12 and still about **half** the size for zoom 13-16

Two conclusions:
- thunderforest is quite far off from the others, especially on lower zooms
- even though jawg.io serves tiles till zoom 16, it doesn't make it bigger (than openmaptiles), each of those z16 tiles is tiny.

**Comment by westnordost** [comment]
After this evaluation, I decided that the best choice for StreetComplete are jawg tiles:
- They are the smallest, which is important for use on mobile devices. That of course also means that they have the fewest features, but the map style of StreetComplete is simple anyway
- They are the most current

JawgMaps also provides the tiles to StreetComplete for free, as you can now read in the readme and in the changelog of the new beta I am compiling right now. Though, the other providers I talked with were also very forthcoming to reach an arrangement and open for feedback. I can recommend all the providers I looked, each having some unique selling points:
- JawgMaps has the most current and lightweight tiles
- Thunderforest has the most detailed tiles with a lot of different information on it and are also very current
- Providers offering MapTiler and StadiaMaps are economically priced and can be considered as more compatible as they build on an open standard (OpenMapTiles)

**Comment by ianthetechie** [comment]
Thanks again for posting your detailed analysis here. This is really interesting stuff, and the feedback is really valuable for helping us improve things like tile sizes.

**Comment by westnordost** [comment]
Well this size-analysis was quite rough, I'd expect the values to very much fluctuate in different areas of the world.

But what I can definitely say from that is that there is room for improvement when handling mid and zoom levels for all four of the services I looked at in one way or another for:
- better simplifying of geometry on lower zooms 
- merging together neighbouring same(ish) landuses and -covers

JawgMaps has an advantage here because they throw many landuses and covers into one pot, for example industrial and residential landuses are both just rehashed to "other" landuses. So if they are the same, they can be merged together for lower zoom levels.
For tile providers who want more detail, this means that it is expedient to group landcovers in a hierarchical manner. For example if `amenity=school` was mapped to something like `layer: landcover, class: populated, populated: residential, residential=education` or similar, on lower zoom levels, these polygons could be merged together with other `populated` polygons.

**Comment by RubenKelevra** [comment]
Thank you for investing so much time in this issue! The new service works great! It makes fun to start the app again, since you don't wait like 10 seconds on a blank screen. :)

**Comment by michaelblyons** [comment]
I'm having trouble installing the Beta overtop my F-Droid provided version.

**Comment by matkoniecz** [comment]
See https://www.chiark.greenend.org.uk/~sgtatham/bugs.html - it is completely unclear what is going wrong. Error message or explanation what exactly is going wrong is necessary (for example "StreetComplete crashes on start" is different from "install fails because I run out of memory" or "F-Droid is down"). Also, it is basically certain that it is a new unrelated issue.

**Comment by Helium314** [comment]
This is because the F-Droid version is signed with F-Droid keys, so you can only update with versions from F-Droid.
If you have root access there are some ways around this, otherwise you need to remove the F-Droid version and then install the Beta.

**Comment by smichel17** [comment]
The debug version can also be installed alongside the release version, if you're up for installing Android Studio on your computer and building the project yourself.

**Comment by michaelblyons** [comment]
21.0 is on F-Droid now. It's so great! Thank you!

**Comment by sinekonata** [comment]
> After this evaluation, I decided that the best choice for StreetComplete are jawg tiles:
> ...

Does this mean that it is now implemented? Is StreetComplete using the tiles that would re-generate daily? In Belgium, they re-generate every 3 days, which is quite good when you see other OSM clients usually have months old tiles.

**Comment by westnordost** [comment]
Yes, since v21.0 (June)
